File: src/ee/storage/DataConflictException.cpp
/* This file is part of VoltDB.
 * Copyright (C) 2008-2011 VoltDB Inc.
 *
 * VoltDB is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * VoltDB is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with VoltDB.  If not, see <http://www.gnu.org/licenses/>.
 */
#include "storage/DataConflictException.h"
#include "storage/persistenttable.h"
#include "storage/table.h"
#include "common/debuglog.h"
#include <cassert>

namespace voltdb {
/*
 * @param table Table that the operation was performed on
 * @param tableId CatalogId of the table that the operation failed on
 * @param beforeTuple before tuple values or null if before operation is Insert
 * @param afterTuple after tuple values or null if after operation is Delete.
 * @param blockerTxnId txn that modified the tuple
 * @param waitingTxnId txn wanted to perform the operation
 */
DataConflictException::DataConflictException(
        PersistentTable *table,
        TableTuple beforeTuple,
        TableTuple afterTuple,
        int64_t blockerTxnId,
        int64_t waitingTxnId) :
            SQLException(
                    SQLException::integrity_data_conflict,
                    "Attempted modification on Escrow-data",
                    voltdb::VOLT_EE_EXCEPTION_TYPE_DATA_CONFLICT),
            m_table(table), m_beforeTuple(beforeTuple), m_afterTuple(afterTuple),
            m_blockerTxnId(blockerTxnId), m_waitingTxnId(waitingTxnId) {
        assert(table);
        assert(!beforeTuple.isNullTuple() || !afterTuple.isNullTuple());
	m_executedPlanNodes = 0;
	m_executedBatch = 0;
	table->printJournalInfo("DCT");
}

DataConflictException::DataConflictException(
        PersistentTable *table,
        int64_t blockerTxnId,
        int64_t waitingTxnId) :
            SQLException(
                    SQLException::integrity_data_conflict,
                    "Attempted modification on Escrow-data",
                    voltdb::VOLT_EE_EXCEPTION_TYPE_DATA_CONFLICT),
            m_table(table), m_blockerTxnId(blockerTxnId), m_waitingTxnId(waitingTxnId) {
        assert(table);
	m_executedPlanNodes = 0;
	m_executedBatch = 0;

}

void DataConflictException::p_serialize(ReferenceSerializeOutput *output) {
    SQLException::p_serialize(output);
    output->writeLong(m_blockerTxnId);
    output->writeLong(m_waitingTxnId);
    output->writeLong(m_executedBatch);
    output->writeLong(m_executedPlanNodes);
    output->writeTextString(m_table->name());

    VOLT_TRACE("WWWG: blocker %jd waiter %jd %jd %jd %s \n", (intmax_t)m_blockerTxnId, (intmax_t)m_waitingTxnId,
    			(intmax_t)m_executedBatch, (intmax_t)m_executedPlanNodes, m_table->name().c_str());
}
DataConflictException::~DataConflictException() {
    // TODO Auto-generated destructor stub
}

}

File: src/frontend/org/voltdb/exceptions/DataConflictException.java
/* This file is part of VoltDB.
 * Copyright (C) 2008-2011 VoltDB Inc.
 *
 * VoltDB is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * VoltDB is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with VoltDB.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.voltdb.exceptions;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.util.logging.Logger;

import org.voltdb.types.ConstraintType;
import org.voltdb.*;
import org.voltdb.messaging.FastDeserializer;
import org.voltdb.messaging.FastSerializer;

/**
 * Exception generated when a constraint is violated. Contains more information then a SQLException.
 *
 */
public class DataConflictException extends SQLException {

    protected static final Logger LOG = Logger.getLogger(DataConflictException.class.getName());

    /**
     * Constructor for deserializing an exception returned from the EE.
     * @param exceptionBuffer
     */
    public DataConflictException(ByteBuffer exceptionBuffer) {
    	    	
        super(exceptionBuffer);
        
        blockerTxnId = exceptionBuffer.getLong();
        waitingTxnId = exceptionBuffer.getLong();
        executedBatch = exceptionBuffer.getLong();
        executedPlanNodes = exceptionBuffer.getLong();
        try {
            tableName = FastDeserializer.readString(exceptionBuffer);
        }
        catch (IOException e) {
            // implies that the EE created an invalid constraint
            // failure, which would be a corruption/defect.
            VoltDB.crashVoltDB();
        }
        
        if (exceptionBuffer.hasRemaining()) {
            int tableSize = exceptionBuffer.getInt();
            buffer = ByteBuffer.allocate(tableSize);
            //Copy the exception details.
            exceptionBuffer.get(buffer.array());
        } else {
            buffer = ByteBuffer.allocate(0);
        }
    }
    
    public long getBlockingTxnId() {
    	return blockerTxnId;
    }
     
    public long getWaitingTxnId() {
    	return waitingTxnId;
    }
    
    public long getExecutedBatch() {
    	return executedBatch;
    }
    
    public long getExecutedPlanNodes() {
    	return executedPlanNodes;
    }

    public String getTableName() {
        return tableName;
    }

    /**
     * Get the table containing the involved tuples (target and source) that caused the constraint violation. The table
     * is lazy deserialized for performance.
     */
    public VoltTable getTuples() {
        if (table != null) {
            return table;
        }

        table = PrivateVoltTableFactory.createUninitializedVoltTable();
        try {
            table.readExternal(new FastDeserializer(buffer));
        } catch (IOException e) {
            LOG.severe(e.toString());
            return null;
        }

        return table;
    }

    /**
     * ByteBuffer containing a serialized copy of a table with the tuples involved in the constraint violation
     */
    private final ByteBuffer buffer;

    /**
     * Type of constraint violation that caused this exception to be thrown
     */
    private final long blockerTxnId;

    private final long waitingTxnId;
    private final long executedBatch;
    private final long executedPlanNodes;
    private String tableName = null;

    /**
     * Lazy deserialized copy of a table with the tuples involved in the constraint violation
     */
    private VoltTable table = null;

    @Override
    public String getMessage() {
        if (buffer.capacity() == 0) {
            return super.getMessage();
        } else {
            StringBuilder sb = new StringBuilder(super.getMessage());
            sb.append('\n');
            sb.append("Blocker Txn ID ");
            sb.append(blockerTxnId);
            
            sb.append("Waiting Txn ID ");
            sb.append(waitingTxnId);
            sb.append("ExecutedPlanNodes ");
            sb.append(executedPlanNodes);
            sb.append(", Table CatalogId ");
            sb.append(tableName);
            sb.append('\n');
            sb.append(getTuples().toString());

            return sb.toString();
        }
    }
    /**
     * Retrieve a string representation of the exception. If this is a rethrown HSQLDB exception it returns the enclosed exceptions
     * string. Otherwise a new string is generated with the details of the constraint violation.
     */
    @Override
    public String toString() {
        return getMessage();
    }

    @Override
    protected SerializableExceptions getExceptionType() {
        return SerializableExceptions.DataConflictException;
    }

    @Override
    protected int p_getSerializedSize() {
        // ... + 8 + string prefix + string length + ...
        return super.p_getSerializedSize() + 8 + 8 + buffer.capacity();
    }

    @Override
    protected void p_serializeToBuffer(ByteBuffer b) throws IOException {
        super.p_serializeToBuffer(b);
        b.putLong(blockerTxnId);
        b.putLong(waitingTxnId);
        b.putLong(executedPlanNodes);
        b.putInt(buffer.capacity());
        buffer.rewind();
        b.put(buffer);
    }

    /**
     *
     */
    private static final long serialVersionUID = 1L;

}
